File: 79-word-search/word-search.java
class Solution {
    boolean result = false;
    void helper(char[][] board, String word, int i, int j, int start){
        if(start >= word.length()){
            result = true;
            return;
        }
        if(i >= board.length || j >= board[0].length || i < 0 || j < 0 || board[i][j] != word.charAt(start)){
            return;
        }
        char temp = board[i][j];
        board[i][j] = '*';
        helper(board, word, i + 1, j, start + 1);
        helper(board, word, i - 1, j, start + 1);
        helper(board, word, i, j + 1, start + 1);
        helper(board, word, i, j - 1, start + 1);
        board[i][j] = temp;


    }
    public boolean exist(char[][] board, String word) {
        for(int i = 0; i < board.length; i++){
            for(int j = 0; j < board[0].length; j++){
                if(board[i][j] == word.charAt(0)){
                    helper(board, word, i, j, 0);
                }
                if(result){
                    return true;
                }
            }
        }
        return false;
        
    }
}

File: 79-word-search/word-search.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        result = False
        def helper(board, i, j, currIndex, word):
            nonlocal result
            if i < 0 or j < 0 or i >= len(board) or j >= len(board[0]) or board[i][j] != word[currIndex]:
                return
            if currIndex == len(word) - 1:
                result = True
                return
            currIndex = currIndex + 1
            temp = board[i][j]
            board[i][j] = '*'
            helper(board, i + 1, j, currIndex, word)
            helper(board, i - 1, j, currIndex, word)
            helper(board, i, j + 1, currIndex, word)
            helper(board, i, j - 1, currIndex, word)
            board[i][j] = temp
        for i in range(len(board)):
            for j in range(len(board[0])):
                if board[i][j] == word[0]:
                    helper(board, i, j, 0, word)
        return result